propolis-server-lib/tests/integration_tests.rs

```diff
@@ -0,0 +1,235 @@
+mod artifacts;
+
+use artifacts::{setup, TEST_BOOTROM, TEST_IMAGE};
+
+fn initialize_log() -> Logger {
+    let config_logging =
+        ConfigLogging::StderrTerminal { level: ConfigLoggingLevel::Info };
+    config_logging.to_logger("propolis-server-integration-test").unwrap()
+}
+
+async fn initialize_server(log: Logger) -> HttpServer<server::Context> {
+    setup().await;
+
+    let mut block_options = BTreeMap::new();
+    block_options.insert(
+        "disk".to_string(),
+        toml::value::Value::String(
+            TEST_IMAGE.path().as_path().to_str().unwrap().to_string(),
+        ),
+    );
+    block_options.insert(
+        "pci-path".to_string(),
+        toml::value::Value::String("0.4.0".to_string()),
+    );
+    let block = Device {
+        driver: "pci-virtio-block".to_string(),
+        options: block_options,
+    };
+    let mut devices = BTreeMap::new();
+    devices.insert("block0".to_string(), block);
+
+    let config = Config::new(TEST_BOOTROM.path(), devices);
+    let context = server::Context::new(config);
+
+    let config_dropshot = ConfigDropshot {
+        bind_address: "127.0.0.1:0".parse().unwrap(),
+        ..Default::default()
+    };
+    HttpServerStarter::new(&config_dropshot, server::api(), context, &log)
+        .unwrap()
+        .start()
+}
+
+// Create a new instance.
+//
+// NOTE: Many of these values are placeholders, and can be replaced
+// when we have "real" integration of UUID-based images / bootroms.
+fn create_ensure_request(
+    id: Uuid,
+) -> propolis_client::api::InstanceEnsureRequest {
+    propolis_client::api::InstanceEnsureRequest {
+        properties: propolis_client::api::InstanceProperties {
+            id,
+            name: "test-instance".to_string(),
+            description: "it's a test instance".to_string(),
+            image_id: Uuid::new_v4(),
+            bootrom_id: Uuid::new_v4(),
+            memory: 256,
+            vcpus: 2,
+        },
+    }
+}
+
+// Utility wrapper for monitoring an instance's state until it reaches a
+// requested target.
+async fn wait_until_state(
+    client: &Client,
+    id: Uuid,
+    state: InstanceState,
+) -> Result<()> {
+    let mut gen = 0;
+    loop {
+        let monitor_response = client.instance_state_monitor(id, gen).await?;
+        if monitor_response.gen < gen {
+            bail!(
+                "Gen should be increasing: (requested {}, saw {})",
+                gen,
+                monitor_response.gen
+            );
+        }
+        eprintln!(
+            "Monitor response: {:?} (waiting for {:?})",
+            monitor_response, state
+        );
+        if monitor_response.state == state {
+            return Ok(());
+        }
+        gen = monitor_response.gen + 1;
+    }
+}
+
+#[tokio::test]
+async fn test_uninitialized_server() {
+    let log = initialize_log();
+    let server = initialize_server(log.clone()).await;
+
+    let client = Client::new(server.local_addr(), log);
+    assert!(matches!(
```

Comment:
Since we're already using nightly, you can collapse these to just `assert_matches!`

Reply:
The use of nightly is mainly to fulfill the requirements of `usdt`.  I think we should constrain our use of nightly features as much as possible (so we can get back on stable ASAP).

Reply:
I'm gonna defer making this change for the reasons Patrick mentioned, but TIL about `assert_matches`.

(Matches is already so useful, definitely interested in making it part of my idiomatic rust toolkit)